File: leetcode/1-100/75-sort-colors/bench/sort_colors.c
/*
 * Benchmark workload — Sort Colors (LeetCode #75).
 * C mirror of bench/sort_colors.kara. Dutch National Flag one-pass sort over an
 * int64_t buffer malloc'd ONCE (n=500) and reused: each of K=200,000 iterations
 * refills it in place with a k-dependent {0,1,2} pattern, sorts in place, and
 * folds the result into a rolling polynomial hash. The measured work is the
 * sort's data-dependent branches and swaps, not allocation.
 * See ../README.md § Benchmarks.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

static void sort_colors(int64_t *a, int64_t n) {
    if (n == 0) return;
    int64_t low = 0, mid = 0, high = n - 1;
    while (mid <= high) {
        if (a[mid] == 0) {
            int64_t t = a[low]; a[low] = a[mid]; a[mid] = t;
            low++;
            mid++;
        } else if (a[mid] == 1) {
            mid++;
        } else {
            int64_t t = a[mid]; a[mid] = a[high]; a[high] = t;
            high--;
        }
    }
}

int main(void) {
    const int64_t n = 500, total = 200000, modulus = 1000000007;
    int64_t *a = (int64_t *)malloc(sizeof(int64_t) * (size_t)n);

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        for (int64_t j = 0; j < n; j++)
            a[j] = (j * 7 + k * 13) % 3;
        sort_colors(a, n);
        for (int64_t j = 0; j < n; j++)
            acc = (acc * 131 + a[j]) % modulus;
    }
    printf("%lld\n", (long long)acc);

    free(a);
    return 0;
}

File: leetcode/1-100/75-sort-colors/bench/sort_colors.rs
// Benchmark workload — Sort Colors (LeetCode #75), seq lane.
// Rust single-threaded mirror of bench/sort_colors.kara. A batch of K=2000
// independent Dutch National Flag sorts of n=59999 {0,1,2} arrays (length not a
// multiple of 3, so the sorted result depends on the seed), each hashed, combined
// through a plain associative sum. This is the single-threaded baseline; the
// rayon variant (rayon/) is the hand-tuned parallel comparator for Kāra's
// auto-par. See ../README.md § Benchmarks.

fn sort_and_hash(seed: i64) -> i64 {
    let n: i64 = 59999;
    let mut a: Vec<i64> = Vec::new();
    for j in 0..n {
        a.push((j * 7 + seed) % 3);
    }

    let mut low: usize = 0;
    let mut mid: usize = 0;
    let mut high: i64 = n - 1;
    while mid as i64 <= high {
        if a[mid] == 0 {
            a.swap(low, mid);
            low += 1;
            mid += 1;
        } else if a[mid] == 1 {
            mid += 1;
        } else {
            a.swap(mid, high as usize);
            high -= 1;
        }
    }

    let mut acc: i64 = 0;
    for j in 0..n as usize {
        acc = (acc * 131 + a[j]) % 1_000_000_007;
    }
    acc
}

fn main() {
    const TOTAL: i64 = 2000;
    let mut sum: i64 = 0;
    for i in 0..TOTAL {
        sum += sort_and_hash(i);
    }
    println!("{}", sum);
}
